sqlite3/lib/src/ffi/impl/database.dart

```diff
@@ -434,4 +434,112 @@ class DatabaseImpl extends Database {
       _finalizable._statements.remove(stmt._finalizable);
     }
   }
+
+  /// check if this is a in-memory database
+  @visibleForTesting
+  bool isInMemory() {
+    final zDbName = allocateZeroTerminated('main');
+    final pFileName = _bindings.sqlite3_db_filename(_handle, zDbName);
+
+    zDbName.free();
+
+    return pFileName.isNullPointer || pFileName.readString().isEmpty;
+  }
+
+  /// Ported from https://www.sqlite.org/backup.html Example 1
+  void _loadOrSaveInMemoryDatabase(Database other, bool isSave) {
+    final fromDatabase = isSave ? this : other;
+    final toDatabase = isSave ? other : this;
+
+    final zDestDb = allocateZeroTerminated('main');
+    final zSrcDb = allocateZeroTerminated('main');
+
+    final pBackup = _bindings.sqlite3_backup_init(
+        toDatabase.handle.cast(), zDestDb, fromDatabase.handle.cast(), zSrcDb);
+
+    if (!pBackup.isNullPointer) {
+      _bindings.sqlite3_backup_step(pBackup, -1);
+      _bindings.sqlite3_backup_finish(pBackup);
+    }
+
+    final extendedErrorCode =
+        _bindings.sqlite3_extended_errcode(toDatabase.handle.cast());
+    final errorCode = extendedErrorCode & 0xFF;
+
+    zDestDb.free();
+    zSrcDb.free();
+
+    if (errorCode != SqlError.SQLITE_OK) {
+      if (errorCode != SqlError.SQLITE_OK) {
+        throw createExceptionFromExtendedCode(
+            _bindings, _handle, errorCode, extendedErrorCode);
+      }
+    }
+  }
+
+  /// Ported from https://www.sqlite.org/backup.html Example 2
+  void _backupDatabase(
+      Database toDatabase, BackupProgressCallback? progressCallback) {
+    final zDestDb = allocateZeroTerminated('main');
+    final zSrcDb = allocateZeroTerminated('main');
+
+    final pBackup = _bindings.sqlite3_backup_init(
+        toDatabase.handle.cast(), zDestDb, _handle, zSrcDb);
+
+    int returnCode;
+    if (!pBackup.isNullPointer) {
+      do {
+        returnCode = _bindings.sqlite3_backup_step(pBackup, 5);
+
+        final remaining = _bindings.sqlite3_backup_remaining(pBackup);
+        final count = _bindings.sqlite3_backup_pagecount(pBackup);
+
+        progressCallback?.call(100 * (count - remaining) / count);
+
+        if (returnCode == SqlError.SQLITE_OK ||
+            returnCode == SqlError.SQLITE_BUSY ||
+            returnCode == SqlError.SQLITE_LOCKED) {
+          _bindings.sqlite3_sleep(250);
```

Comment:
Do you know whether this needs to be synchronous or if we could resume backups across asynchronous suspensions?

If we can get away with it, I think I'd prefer an API that is essentially `Stream<double> backup(Database toDatabase)` to report progress. Whenever we're entering a locked state, can we just do `await Future.delayed(const Duration(milliseconds: 250))` instead? We can't avoid blocking due to IO, but I'd prefer not blocking Dart threads in `sleep` if possible.

Reply:
I also like the `Stream` much more. Done!
I checked the sqlite source and we can replace the sleep it with an delayed future. Also done!